File: src/test/resources/FindMiddleNode.java
import java.util.*;
import org.algohub.engine.type.*;

public class Solution {
    /**
     * @param head Linked List head
     * @return  Middle node
     */
    public LinkedListNode<Integer> findMiddleNode(LinkedListNode<Integer> head) {
        if (head == null) return null;
        if (head.next == null) return head;

        LinkedListNode slow = head;
        LinkedListNode fast = head.next;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }
}

File: src/test/resources/FindMiddleNode.rb
class Solution
  def findMiddleNode(head)
    return unless head
    return head unless head.next

    slow = head
    fast = head.next

    while fast && fast.next
      slow = slow.next
      fast = fast.next.next
    end

    if fast == nil then
        slow
    else
        slow.next
    end
  end
end
